nest/core/protocols.py

```diff
@@ -0,0 +1,464 @@
+        """
+        ...
+
+    def add_middleware(
+        self,
+        middleware_cls: Any,
+        **options: Any,
+    ) -> None:
+        """
+        Add a middleware class to the application, with config options.
+        """
+        ...
+
+    def run(self, host: str = "127.0.0.1", port: int = 8000) -> None:
+        """
+        Blockingly run the HTTP server on the given host/port.
+        In production, you might prefer to return an ASGI app
+        and let an external server (e.g., gunicorn) run it.
+        """
+        ...
+
+    async def startup(self) -> None:
+        """
+        Optional: If the framework has an 'on_startup' event, run it.
+        """
+        ...
+
+    async def shutdown(self) -> None:
+        """
+        Optional: If the framework has an 'on_shutdown' event, run it.
+        """
+        ...
+
+    def register_routes(self, container: Container) -> None:
+        """
+        Register multiple routes at once.
+        """
+        ...
+
+
+@runtime_checkable
+class CLIAdapterProtocol(Protocol):
+    """
+    High-level interface for an CLI adapter.
+    The PyNest system can call these methods to:
+      - create and manage the main application object
+      - run the cli app
+      - register commands into the cli
+    """
+
+    def create_app(self, **kwargs: Any) -> Any:
+        """
+        Create and store the main CLI application object.
+        **kwargs** can pass parameters
+        Returns the native app object (like FastAPI instance).
+        """
+        ...
+
+    def get_router(self) -> RouterProtocol:
+        """
+        Return a RouterProtocol for the main router (or a sub-router).
+        """
+        ...
+
+    def add_middleware(
+        self,
+        middleware_cls: Any,
+        **options: Any,
+    ) -> None:
+        """
+        Add a middleware class to the application, with config options.
+        """
+        ...
+
+    def run(self, host: str = "127.0.0.1", port: int = 8000) -> None:
+        """
+        Blockingly run the HTTP server on the given host/port.
+        In production, you might prefer to return an ASGI app
+        and let an external server (e.g., gunicorn) run it.
+        """
+        ...
+
+    async def startup(self) -> None:
+        """
+        Optional: If the framework has an 'on_startup' event, run it.
+        """
+        ...
+
+    async def shutdown(self) -> None:
+        """
+        Optional: If the framework has an 'on_shutdown' event, run it.
+        """
+        ...
+
+    def register_routes(self, container: Container) -> None:
+        """
+        Register multiple routes at once.
+        """
+        ...
+
+
+
+
+###############################################################################
+# 3. PARAMETER EXTRACTION (NEST-JS LIKE)
+###############################################################################
+
+T = TypeVar("T")
+
+
+class RequestAttribute:
```

Comment:
We should talk , can use __new__ to register new Param request and reuse it. 
__init_subclass__ -> for register 
__new__ -> for casting